### PR TITLE
feat!: add `length_width` to `schema::Definition::Sequence`

### LINKED ISSUE #211: BorshSchema vs custom serialisation
Say I’d like to use varint in borsh. Or have a custom SmallVec type which is encoded with 8-bit length rather than 32-bit length.

This is easy enough to do by implementing custom BorshSerialize and BorshDeserialize. However, BorshSchema becomes an issue. Varint could be modelled as a nested enum with 256 variants. Similarly SmallVec could be modeled as an enum with 256 variants each being an array. That’s hardly a clean solution though.

Do you guys have any thoughts on that?

### PR DESCRIPTION
**Pull request by mina86**
Add Definition::Sequence::length_width field which allows specifying
the width of the length tag for dynamically-sized arrays.  This allows
defining custom schema where length is, say, one-byte.

Furthermore, support length_width being zero which indicates untagged
sequences.  Use that as a new way to describe a fixed-length arrays
thus making Definition::Array no longer necessary.

Lastly, zero length_width with length range which is more than just
one value allows to document a custom encoding which cannot be fully
specified with the schema.

resolves #211 (`VarInt` and `SmallVec<T>`)

### CODE REVIEW

**Comment by dj8yfo** [comment]
Duplicate of #228

File: borsh/src/schema/container_ext/validate.rs [hunk 1]
```diff
@@ -78,9 +92,7 @@ fn validate_impl<'a>(
             tag_width,
             variants,
         } => {
-            if *tag_width > 8 {
-                return Err(Error::TagTooWide(declaration.to_string()));
-            }
+            check_tag_width(declaration, *tag_width, variants.len() as u64)?;
```
Comment (dj8yfo) [inline borsh/src/schema/container_ext/validate.rs:95]: using `variants.len()` is incorrect here, as it should be `variants.find_max_discriminant()` of `i64` type, and info about discriminants is not collected into `variants`

**Comment by dj8yfo** [comment]
@mina86, can you make #229 identical in content (with respect to code and tests, modify doc freely) to #228? 
We'll then merge #229.

File: borsh/src/schema/container_ext/validate.rs [hunk 2]
```diff
@@ -53,14 +66,15 @@ fn validate_impl<'a>(
     stack.push(declaration);
     match definition {
         Definition::Primitive(_size) => {}
-        Definition::Array { elements, .. } => validate_impl(elements, schema, stack)?,
```
Comment (dj8yfo) [inline borsh/src/schema/container_ext/validate.rs:56]:
```bash
thread 'validate_for_derived_types' panicked at 'assertion failed: `(left == right)`
  left: `Ok(())`,
 right: `Err(ZSTSequence("Array<nil, 300>"))`'
```

**Review by dj8yfo — APPROVED** [review]

File: borsh/src/schema/container_ext/validate.rs [hunk 3]
```diff
@@ -32,23 +32,28 @@ pub enum Error {
     /// Declared tag width is too small.  Tags must be large enough to represent
     /// possible length of sequence.
     TagTooNarrow(Declaration),
+    /// only 0, 1, 2, 4 and 8 bytes long enum tags and sequences' `length_width` are alowed
+    TagNotPowerOfTwo(Declaration),
```
Comment (mina86) [inline borsh/src/schema/container_ext/validate.rs:36]: Again, why?

File: borsh/src/schema/container_ext/validate.rs [hunk 4]
```diff
@@ -92,13 +107,9 @@ fn validate_impl<'a>(
...
-            // TODO(#230): variants.len() is actually incorrect here.  Since
-            // arbitrary discriminant can be used for variants, an actual
-            // largest discriminant may be larger than the number of variants.
-            // For the time being we’re using variants.len() as a proxy but
-            // eventually variant discriminants should be added to the schema.
-            let max_discriminant = variants.len().saturating_sub(1) as u64;
-            check_tag_width(declaration, *tag_width, max_discriminant)?;
+            if *tag_width > U64_LEN {
+                return Err(Error::TagTooWide(declaration.to_string()));
+            }
```
Comment (mina86) [inline borsh/src/schema/container_ext/validate.rs:112]: So `length_width: 3` is invalid (for whatever reason), but `tag_width: 3` is valid?

**Pull request merged**